### PR TITLE
Technical assesment done by Abdurrahman AKKUŞ

### PR DESCRIPTION
**Pull request by abdurrahman-akkus**
Technical assesment done by Abdurrahman AKKUŞ

### CODE REVIEW

File: nb-configuration.xml [hunk 1]
```diff
@@ -0,0 +1,31 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (HasseNasse) [inline nb-configuration.xml]: IDE specific config files should not be committed

**Comment by HasseNasse** [comment]
Can we please start off by cleaning some stuff away, please remove the IDE specific config files and target folder. @akkus12345

File: nb-configuration.xml (hunk 1 above)
Comment (abdurrahman-akkus) [reply nb-configuration.xml]: Unwanted files removed from project. They added to .gitignore.

**Comment by abdurrahman-akkus** [comment]
> Technical assesment done by Abdurrahman AKKUŞ

**Comment by HasseNasse** [comment]
Thanks for your submission Abdurrahman!
I have some questions out of the bat:
- Can you please detail your EC2 setup, did you use any IaC? Are you using docker on top of the EC2 instance?
- As the growing pandemic might entail an exponential scale-up in callers towards this service, because more and more govermental and private health institutes will start polling and/or pushing data to the service. Did you consider the scalability aspects of the app + infrastructure setup. (e.g deploying out the app as a ECS container instance / serverless).
- Did you consider what the cost most cost effective solution could've been? If so, please detail what improvements can be made to allow for higher scalability (both in terms of performance and cost).

**FYI:**
No need to do any changes at this point, We are simply asking you to contemplate on the solution and consider what improvements could've been done.

File: src/main/java/com/akkus/RestDemoApplication.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+@SpringBootApplication
+@EnableScheduling
+@EnableAutoConfiguration
+@ComponentScan("com.akkus.*")
+public class RestDemoApplication {
+
+    /**
+     *
+     */
+    public final static Logger logger = getLogger(RestDemoApplication.class);
```
Comment (HasseNasse) [inline src/main/java/com/akkus/RestDemoApplication.java:26]: What is the purpose of this line?

File: src/main/java/com/akkus/RestDemoApplication.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+    /**
+     *
+     * @param args
+     */
+    public static void main(String[] args) {
+        run(RestDemoApplication.class, args);
+    }
+    private static final java.util.logging.Logger LOG = java.util.logging.Logger.getLogger(RestDemoApplication.class.getName());
```
Comment (HasseNasse) [inline src/main/java/com/akkus/RestDemoApplication.java:35]: What is the purpose of this line?

File: src/main/java/com/akkus/ServletInitializer.java [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+    /**
+     *
+     * @param application
+     * @return
+     */
+    @Override
+    protected SpringApplicationBuilder configure(SpringApplicationBuilder application) {
+        return application.sources(RestDemoApplication.class);
+    }
+    private static final Logger LOG = Logger.getLogger(ServletInitializer.class.getName());
```
Comment (HasseNasse) [inline src/main/java/com/akkus/ServletInitializer.java:22]: What is the purpose of this line?

File: src/main/java/com/akkus/ServletInitializer.java [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+import java.util.logging.Logger;
+import org.springframework.boot.builder.SpringApplicationBuilder;
+import org.springframework.boot.web.servlet.support.SpringBootServletInitializer;
+
+/**
+ *
+ * @author akkus
+ */
+public class ServletInitializer extends SpringBootServletInitializer {
```
Comment (HasseNasse) [inline src/main/java/com/akkus/ServletInitializer.java:11]: Are we using servlets? I can only see a @RestControllerMapping

File: src/main/java/com/akkus/gen/ISuperApi.java [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+import java.util.List;
+import org.springframework.http.ResponseEntity;
+
+/**
+ *
+ * @author dijital
+ * @param <T>
+ */
+// ISuperEntity interface de olsa generic constraint kavramı için extends ile kullanılır!!!
```
Comment (HasseNasse) [inline src/main/java/com/akkus/gen/ISuperApi.java:16]: ?

File: src/main/java/com/akkus/gen/VocidService.java [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+import org.hibernate.SessionFactory;
+import org.hibernate.cfg.Configuration;
+import org.springframework.stereotype.Service;
+
+/**
+ *
+ * @author Abdurrahman AKKUŞ
+ */
+@Service
+public class VocidService extends ISuperService<ISuperEntity> {
```
Comment (HasseNasse) [inline src/main/java/com/akkus/gen/VocidService.java:19]: I would've liked a clearer seperation between the repository and service layers.

File: src/main/java/com/akkus/gen/VocidEntity.java [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+ * @author Abdurrahman AKKUŞ 
+ */
+@Data
+@Entity
+@Table(name="vocid")
+public class VocidEntity implements ISuperEntity {
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    @Column(name="id")
```
Comment (HasseNasse) [inline src/main/java/com/akkus/gen/VocidEntity.java:26]: `@Column(name="id")`
is not needed when the field names are the same on the entity.

File: src/main/resources/application.properties [hunk 9]
```diff
@@ -0,0 +1,10 @@
+#server.servlet.contextPath=/
+spring.servlet.multipart.max-file-size= 1024000000
+spring.servlet.multipart.max-request-size= 1024000000
+spring.servlet.multipart.location= /tmp
+#public ve private yanl\u0131\u015f yerlere yaz\u0131lm\u0131\u015f ancak b\u0131rak \u00f6yle kals\u0131n de\u011filse sistem hata veriyor.
```
Comment (HasseNasse) [inline src/main/resources/application.properties:5]: What is this?

**Review by HasseNasse — COMMENTED** [review]
Nicely written brother!

File: src/main/java/com/akkus/RestDemoApplication.java (hunk 2 above)
Comment (abdurrahman-akkus) [reply src/main/java/com/akkus/RestDemoApplication.java:26]: Actually I used one of my old project as template.I have missed to delete this part.

File: src/main/java/com/akkus/RestDemoApplication.java (hunk 3 above)
Comment (abdurrahman-akkus) [reply src/main/java/com/akkus/RestDemoApplication.java:35]: Actually I used one of my old project as template.I have missed to delete this part.

File: src/main/java/com/akkus/ServletInitializer.java (hunk 4 above)
Comment (abdurrahman-akkus) [reply src/main/java/com/akkus/ServletInitializer.java:22]: Actually I used one of my old project as template.I have missed to delete this part.

**Comment by abdurrahman-akkus** [comment]
> Thanks for your submission Abdurrahman! I have some questions out of the bat:
> 
>     * Can you please detail your EC2 setup, did you use any IaC? Are you using docker on top of the EC2 instance?
> 
>     * As the growing pandemic might entail an exponential scale-up in callers towards this service, because more and more govermental and private health institutes will start polling and/or pushing data to the service. Did you consider the scalability aspects of the app + infrastructure setup. (e.g deploying out the app as a ECS container instance / serverless).
> 
>     * Did you consider what the cost most cost effective solution could've been? If so, please detail what improvements can be made to allow for higher scalability (both in terms of performance and cost).
> 
> 
> **FYI:** No need to do any changes at this point, We are simply asking you to contemplate on the solution and consider what improvements could've been done.

I didn't use IaC(I yet learned) or docker. I initiated a server. Then sent tomcat to the server via scp. I changed user and roles of tomcat. I initiated tomcat service. For javadoc I need a web server, because javadoc was composed of html files. I installed Apache and I put javadoc inside. I installed MySQL and configured it. 

For scale up the project we can spread our API to other servers. To be able to find out which server will be accessed we would need a load balancer. MySQL may not enough. We can use more instances, some of them used for only reading others for only writing. MySQL which is used for reading should be at least a master-slave relationship. Or we can use all instances for both reading and writing. Then we must establish master-master relationship between them.

**Comment by HasseNasse** [comment]
> I didn't use IaC(I yet learned) or docker. I initiated a server. Then sent tomcat to the server via scp. I changed user and roles of tomcat. I initiated tomcat service. For javadoc I need a web server, because javadoc was composed of html files. I installed Apache and I put javadoc inside. I installed MySQL and configured it.

> For scale up the project we can spread our API to other servers. To be able to find out which server will be accessed we would need a load balancer. MySQL may not enough. We can use more instances, some of them used for only reading others for only writing. MySQL which is used for reading should be at least a master-slave relationship. Or we can use all instances for both reading and writing. Then we must establish master-master relationship between them.

Thank you for your reply!

**Pull request closed without merging**